pandas/core/indexes/base.py

```diff
@@ -1278,39 +1278,46 @@ def _set_names(self, values, level=None):
 
     def set_names(self, names, level=None, inplace=False):
         """
-        Set new names on index. Defaults to returning new index.
+        Set Index or MultiIndex name.
+
+        Able to set new names partially and by level.
 
         Parameters
         ----------
-        names : str or sequence
-            name(s) to set
-        level : int, level name, or sequence of int/level names (default None)
-            If the index is a MultiIndex (hierarchical), level(s) to set (None
-            for all levels).  Otherwise level must be None
-        inplace : bool
-            if True, mutates in place
+        names : int, str, list-like, optional
```

Comment:
I don't think we should try to list the possible types here (`int, str`), as in principle an index name can be anything hashable, so not only ints or strings but also timestamps, tuples, ... 
I seem to remember that we had such a discussion during the documentation sprint recently, and we ended up using the generic "label" (and then it can be "label, or list of labels")

BTW, this one is not optional I think?

Reply:
right, it was ended with the generic "label". but @datapythonista said that we should follow this. 
https://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html#section-3-parameters

So I was not sure that using "label" in there is right. which one do you think better?
and yes, it is not optional. it because of my bad misunderstanding. I will delete it.

Reply:
Yeah, sorry, we should update that text

Reply:
Yes, sorry, I meant `object` instead of `str`, this is what we've been using for what I know. We tried to keep them as Python types, but I'm happy using `label`, it may be clearer. But if we use that, I'd prefer `list of label` to `list of labels`. This is consistent with `list of str` or `list of int`, and it will make it easier to parse and detect types automatically, if at some point that is useful.

Reply:
In this case I think `names` expects `label or list of label` if we go for the `label` option. Or `object or list of object` if we prefer the Python type. Am I right?

Reply:
I think `label or list of label` looks better